src/irmin/metrics.mli

```diff
@@ -0,0 +1,48 @@
+(*
+ * Copyright (c) 2022 Etienne Marais <etienne@maiste.fr>
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *)
+
+type origin = ..
+type origin += Origin_default
+type 'a metric = { name : string; uid : int; origin : origin; ty : 'a Repr.t }
+
+type ('a, 'b) data = private {
+  uid : int;
+  name : string;
+  metric : 'a metric;
+  mutable v : 'a;
+  mutable storage : 'b;
+}
+
+val value : ('a, 'b) data -> 'a
+val storage : ('a, 'b) data -> 'b
+
+type ('a, 'b) kind =
+  | Ref
+  | Value of ('a -> 'a)
+  | Storage of ('a -> 'b -> 'b)
```

Comment:
I'm not sure to understand what these types mean :-) Might be useful to add comments/examples here?

Reply:
Yes, sure :smile: For now, it's just a draft so that @Ngoguey42 can look at it, and we have a place to discuss it.
But the plan is to improve the interface and add some documentation as soon as we are sure about it.

Reply:
I updated the documentation.

Reply:
A good way to design a good API is actually to try to explain it with documentation :-) ie. always start by a `.mli`, give a few hints on the general idea behind the API and give a few explanations on how users are supposed to use it. Writing a few examples is always a good way to find the right function and types names and the proper API to compose them.

Reply:
(I am still not totally sure how to use the current API bu just looking at the mli - some the names do not makes much sense - for instance, why `Ref`?)

Reply:
Thanks for your explanation! I add this to my "Good Practises Guide" :smile: 
Moreover, I'll add some examples / hints at the beginning of the code as you suggest.

The `kind` type is supposed to describe three types of behaviours:
- The first one does nothing by itself
- The second one applies the function to update the `value` but not the storage
- The last one alters the `storage` and not the value
Of course, it's still possible to add more with other behaviours.